specs/DesignDiscussions/DotNetTemplate.md

```diff
@@ -0,0 +1,348 @@
+When a data context node is encountered, it call a base visit method with the node, then the method resolved into VisitTemplateData method.
+The method, creates a data context object with a field that indicates Array if set. The data context object parse the value in the pair of the data context node into JArray object if it's array, otherwise into JObject
+The newly created data context object is set to current context. In current design, it uses a stack for this purpose. When the method returns to the VisitObj method, data context is set, and translation can continue. Please refer to **Figure 4**
+* Do we use first data context we see, or do we have to see all of them?
+
+# Translation
+## Expansion in Basic Form
+This is the most basic form of expansion that involves templated string and dictionary type data context 
+#### property 
+Data is accessed via '.' 
+```json
+{
+    "type": "AdaptiveCard",
+    "body": [
+        {
+            "type": "TextBlock",
+            "text": "Hi {employee.name}! Here's a bit about your org..."
+        }
+    ]
+}
+```
+
+#### array 
+Data is accessed via '[]' with index 
+
+```json
+{
+    "type": "AdaptivCard",
+    "body": [
+        {
+            "type": "TextBlock",
+            "text": "3 of your peers are: {employee.peers[0].name}, {employee.peers[1].name}, {employee.peers[2].name}"
+        }
+    ]
+}
+```
+The distinction in accessors type was important before the use of AEL. It's not important anymore since the job of expansion is delegated to AEL.
+## Parsing Strategy
+Terminal Nodes or leaf nodes contain the templated string, so the default VisitTerminal method is overridden. When the method sees that given parse context is a TEMPLATELITERAL, it tries to expand using AEL & given data context. Since our data context is JObject or JArray type that implement IDictionary, AEL can expand it. When expansion fails, the method indicate that the literal has been expanded, so when the result is returned, the unaltered literal will be included in the result     
+* Including the unaltered literal when the expansion fails allows 2nd pass when data becomes avaialable, should this be default behavior? 
+
+## Expansion with Array Data Context                                                                                                                         
+When data context type is an array and there is an adaptive element that uses the data context. Template lagunage specifies that the adaptive element that hosts the data context repeats its subtree of which root node is iteslf.  
+  
+**Figure 5**  
+![](Drawing1.png)  
+  
+**Figure 6**  
+![](Drawing2.png)   
+```json
+{
+    "type": "AdaptiveCard",
+    "version": "1.0",
+    "body": [
+        {
+            "type": "Container",
+            "items": [
+                {
+                    "type": "TextBlock",
+                    "text": "Header"
+                },
+                {
+                    "$data": [
```

Comment:
>$data [](start = 21, length = 5)

Discussion: This will repeat the fact set. Might make more sense to update the example to update the facts instead.